live-build/misc/upgrade-scripts/upgrade-container

```diff
@@ -273,12 +273,149 @@ function run() {
+			die "'cp $path' failed for '$CONTAINER'"
+	fi
+}
+
+function convert_to_bootfs_cleanup() {
+	for dir in /proc /sys /dev; do
+		umount -R "/var/lib/machines/${CONTAINER}${dir}" ||
+			warn "'umount -R' of '$dir' failed"
+	done
+}
+
+#
+# The purpose of this function is to convert an existing upgrade
+# container (specified by the $CONTAINER global variable) into the next
+# boot filesystem used by the appliance; i.e. after calling this
+# function, the container's root filesystem will be used as the
+# appliance's root filesystem, the next time the appliance boots.
+#
+# This is done by updating the appliance's bootloader (i.e. grub) to
+# point to the container's filesystem, along with setting the mountpoint
+# of the filesystem to be "/" instead of "/var/lib/machines/$CONTAINER".
+# This effectively removes the container, since systemd-nspawn only
+# looks in /var/lib/machines, so it'll no longer find the container
+# after the mountpoint changes.
+#
+function convert_to_bootfs() {
+	trap convert_to_bootfs_cleanup EXIT
+
+	for dir in /proc /sys /dev; do
+		mount --rbind "$dir" "/var/lib/machines/${CONTAINER}${dir}" ||
+			die "'mount --rbind' of '$dir' failed"
+		mount --make-rslave "/var/lib/machines/${CONTAINER}${dir}" ||
+			die "'mount --make-rslave' of '$dir' failed"
+	done
+
+	#
+	# When performing a not-in-place upgrade, the root and delphix
+	# users will not have any password by default. Thus, we need to
+	# explicitly configure the passwords on the new root filesystem.
+	# Here, we ensure the passwords for the root and delphix user on
+	# the new root filesystem match their current values.
+	#
+	migrate_password_for_user delphix
+	migrate_password_for_user root
+
+	#
+	# LX-72 Until we have a proper solution for migrating the
+	# configuration of a given appliance to the new root filesystem
+	# (i.e. any configuration that isn't supplied by a package), we
+	# resort to explicitly migrating some of this configuration
+	# here. The files listed here are dynamically generated and/or
+	# modified by appliance-build.
+	#
+	while read -r file; do
+		migrate_file "$file"
+	done <<-EOF
+		/etc/cloud/cloud.cfg.d/99-delphix-development.cfg
+		/etc/cloud/cloud.cfg.d/99-delphix-internal.cfg
+		/etc/issue
+		/etc/nftables.conf
+		/etc/ssh/sshd_config
+		/etc/systemd/system/delphix-masking.service.d/override.conf
+		/etc/systemd/system/delphix-mgmt.service.d/override.conf
+		/etc/systemd/system/delphix-postgres@.service.d/override.conf
+		/var/opt/delphix/server.conf
+	EOF
+
+	chroot "/var/lib/machines/$CONTAINER" update-grub ||
+		die "'update-grub' failed in '$CONTAINER'"
+
+	for dev in $(get_bootloader_devices); do
+		[[ -e "/dev/$dev" ]] ||
+			die "bootloader device '/dev/$dev' not found"
+
+		[[ -b "/dev/$dev" ]] ||
+			die "bootloader device '/dev/$dev' not block device"
+
+		chroot "/var/lib/machines/$CONTAINER" \
+			grub-install "/dev/$dev" ||
+			die "'grub-install /dev/$dev' failed in '$CONTAINER'"
```

Comment:
Did you figure out how does grub determine which ZFS dataset to get the `/boot/` directory from? Since it needs to mount the right dataset to get `/boot/grub.cfg`, it can't be any entry in `grub.cfg` that defines which dataset needs to be mounted at first.

Reply:
I'm not sure, but I didn't verify running `update-grub` and `grub-install` from the `chroot` works (i.e. the system boots to w/e dataset I used to `chroot`).

The contents of `/boot/grub/grub.cfg` in the `chroot` dataset will be "correct", while the running system's `/boot/grub/grub.cfg` will be "stale".

I don't really know the details about how grub will determine which `grub.cfg` file to use, but in my testing, the filesystem I expected to boot (i.e. the new one) was booted after running these commands.